### PR TITLE
Fix warning when newline in hintpath

### PR DESCRIPTION
**Pull request by wli3**
### Description

If there are newlines or white spaces passed to `Path.GetFullpath`, it will throw on fullframework. [Adding `Path.GetFullpath`](https://github.com/microsoft/msbuild/pull/4414/files) in previous change caused extra warning from Resolve Assembly Reference. This change added "skip when the path is not valid" to being back the "no warning but skip" behavior of 16.1.

### Customer Impact

If the customer has _warning as error_ enabled, a build would be successful in 16.1 but would be fail in the latest version

### Regression?

Yes. by https://github.com/microsoft/msbuild/pull/4414, since 16.2

### Risk

Low risk.

### Test changes in this PR

Additional unit tests and manual end to end tests.

------------------------------

fix https://github.com/microsoft/msbuild/issues/4593

This is a strict fix for the warning problem. Please look at `CannotNormalizePathWithNewLineAndSpace` test. `FileUtilities.NormalizePath` will get garbage output like (`c:\work\test   \r\n c:\work\referenced\file.dll`) when called with newline. So `File.Exists` cannot find the correct file anyway. The fix is only to make it quiet.

It behaves differently in core because Path.GetFullpath no longer throws when there is invalid char
To be discuss

1. Should we make it work? Instead of skip when contains invalid char, should we Trim it instead?
2. Is master the right branch? Do I need ask mode template for 16.3?

### CODE REVIEW

**Comment by nguerrera** [comment]
I don't understand how the repro project I saw in 16.1 would have been quiet, but not resolve to the path. I would have expected a build warning for unresolved reference if the path was not interpreted as trimmed in 2.1. Did it resolve from another resolver than HintPath?

**Comment by nguerrera** [comment]
I definitely don't want to have no warning, but also ignore the hint path. I'm assuming either warn and ignore or don't warn and trim.

**Comment by nguerrera** [comment]
Do you have a binlog, or can you make a binlog for 16.1 building this: https://github.com/microsoft/msbuild/issues/4593#issuecomment-519292872

**Comment by rainersigwald** [comment]
I agree: I'd like to know exactly what caused the 16.1 behavior before we decide how to fix this.

**Comment by nguerrera** [comment]
Oh, I see the difference. The reference ItemSpec is to System.Net.Http so it resolves to target framework and hint path with new lines is treated as just not found which is not a warning if another resolver finds it.

**Comment by nguerrera** [comment]
So this goes back to what I said originally: we should not trim in sdk conflict resolution if rar does not. What a mess.

**Comment by wli3** [comment]
So which way we should we choose?
a. trim RAR hint path so it can find the actual hint path (and it should, some xml formatter will add the new line)
b. keep the change like this, and stop trimming in conflict resolution and quietly skip when there is one

I prefer a. sounds closer to what customer want

**Comment by rainersigwald** [comment]
a is better behavior, but I think since it's late in 16.3, we should just revert to the 16.1 behavior. Then there's no risk of breaking people with incorrect hint paths that weren't being respected because of whitespace.

**Comment by wli3** [comment]
So take this change as is?

I will later change the sdk part.

**Comment by wli3** [comment]
Also do i need ask mode? 16.3 is master?

**Comment by rainersigwald** [comment]
`master` is still 16.3 for the moment, yeah.

File: src/Tasks/AssemblyDependency/HintPathResolver.cs [hunk 1]
```diff
@@ -51,7 +51,7 @@ public override bool Resolve
             out bool userRequestedSpecificFile
         )
         {
-            if (!string.IsNullOrEmpty(hintPath))
+            if (!string.IsNullOrEmpty(hintPath) && !FileUtilities.PathIsInvalid(hintPath))
```
Comment (rainersigwald) [inline src/Tasks/AssemblyDependency/HintPathResolver.cs]: I think this deserves a comment. Probably with a link to the bug to consider trimming HintPath in 17.0.

Comment (wli3) [reply src/Tasks/AssemblyDependency/HintPathResolver.cs]: Added comment, https://github.com/microsoft/msbuild/pull/4598/commits/fc762c4b478fe94d329bf15b907da88e8e2dfb92

and I do need to have ask mode write up

**Comment by wli3** [comment]
get approved from email. merging

**Pull request merged**